components/system-broker/pkg/oauth/token_provider_from_secret.go

```diff
@@ -33,26 +53,68 @@ type credentials struct {
 	tokensEndpoint string
 }
 
-func NewTokenProviderFromSecret(config *Config, httpClient httputils.Client, k8sClient client.Client) (*OAuthTokenProvider, error) {
-	return &OAuthTokenProvider{
+func NewTokenProviderFromSecret(config *Config, httpClient httputils.Client, tokenTimeout time.Duration, k8sClientConstructor func(time.Duration) (client.Client, error)) (*TokenProviderFromSecret, error) {
+	k8sClient, err := k8sClientConstructor(config.WaitKubeMapperTimeout)
+	if err != nil {
+		return nil, err
+	}
+
+	return &TokenProviderFromSecret{
 		httpClient:        httpClient,
 		k8sClient:         k8sClient,
 		waitSecretTimeout: config.WaitSecretTimeout,
 		secretName:        config.SecretName,
 		secretNamespace:   config.SecretNamespace,
+
+		token:        httputils.Token{},
+		tokenTimeout: tokenTimeout,
+		lock:         sync.RWMutex{},
 	}, nil
 }
 
-func (c *OAuthTokenProvider) GetAuthorizationToken(ctx context.Context) (httputils.Token, error) {
+func (c *TokenProviderFromSecret) Name() string {
+	return "TokenProviderFromSecret"
+}
+
+func (c *TokenProviderFromSecret) Matches(ctx context.Context) bool {
+	if _, err := getBearerToken(ctx); err != nil {
+		log.C(ctx).WithError(err).Warn("while obtaining bearer token")
+		return true
+	}
+
+	return false
+}
+
+func (c *TokenProviderFromSecret) GetAuthorizationToken(ctx context.Context) (httputils.Token, error) {
+	if c.validToken() {
```

Comment:
Guys, we had a discussion here about some potential race conditions. We would love to leave your opinion here.

Reply:
We've resolved the synchronization dilemma. You can have a look.